cmake/SetupPch.cmake

```diff
@@ -0,0 +1,102 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+option(
+  USE_PCH
+  "Use precompiled headers for STL, Blaze, and Brigand"
+  ON
+  )
+
+if (USE_PCH)
+  # We store the header to precompile in ${CMAKE_SOURCE_DIR}/tools so
+  # that it cannot accidentally be included incorrectly anywhere since
+  # ${CMAKE_SOURCE_DIR}/tools is not in the include list.
+  # We copy it to the build directory and include it from there because
+  # GCC's precompiled headers require the hpp and the precompiled header
+  # be in the same directory.
+  set(PCH_PATH "${CMAKE_BINARY_DIR}/SpectrePch.hpp")
+  configure_file(
+    ${CMAKE_SOURCE_DIR}/tools/SpectrePch.hpp
+    ${CMAKE_BINARY_DIR}/SpectrePch.hpp
+    )
+
+  # The compiler flags need to be turned into a CMake list
+  if ("${CMAKE_BUILD_TYPE}" STREQUAL "Debug")
+    string(REPLACE " " ";" PCH_COMPILE_FLAGS
+      "${CMAKE_CXX_FLAGS} ${CMAKE_CXX_FLAGS_DEBUG}"
+      )
+  elseif("${CMAKE_BUILD_TYPE}" STREQUAL "None")
+    string(REPLACE " " ";" PCH_COMPILE_FLAGS "${CMAKE_CXX_FLAGS}")
+  else()
+    string(REPLACE " " ";" PCH_COMPILE_FLAGS
+      "${CMAKE_CXX_FLAGS} ${CMAKE_CXX_FLAGS_RELEASE}"
+      )
+  endif()
+
+  add_custom_command(
+    OUTPUT ${PCH_PATH}.gch
+    COMMAND ${CMAKE_CXX_COMPILER}
+    ARGS
+    -std=c++14
+    ${PCH_COMPILE_FLAGS}
+    -I${CMAKE_SOURCE_DIR}/src
+    -isystem${BLAZE_INCLUDE_DIR}
+    -isystem${BRIGAND_INCLUDE_DIR}
```

Comment:
The problem is these two lines (blaze and brigand).  Their include directories are /usr/include on my system, so these options move /usr/include to the front of the system include search list and then the C++ includes can't find the C ones anymore.  gcc ignores `-I` directives naming system includes to prevent this sort of thing, but apparently not `-isystem`.

Reply:
I wonder how CMake then adds the flags so this doesn't happen when we're normally compiling files. Do have any suggests on how to resolve this? Changing to `-I` will likely result in a spew of warnings from the header files, unfortunately

Reply:
Okay, since we completely control the invocation could we just append the include dirs to CPATH in the call? That might be the easiest way to preserve path ordering

Reply:
Are you saying that if the found include directory is found in the CMAKE_<LANG>... then it just doesn't add the `-isystem`? Because that's also easy to respect

Reply:
Okay, I've pushed a fixup commit to try and fix the issue. Let me know if it works!